File: extensions/shared/src/main/java/app/revanced/extension/shared/patches/client/AppClient.java
package app.revanced.extension.shared.patches.client;

import static app.revanced.extension.shared.utils.ResourceUtils.getString;

import android.os.Build;

import androidx.annotation.Nullable;

public class AppClient {
    
    // WEB
    private static final String CLIENT_VERSION_WEB = "2.20240726.00.00";
    private static final String DEVICE_MODEL_WEB = "Surface Book 3";
    private static final String OS_NAME_WEB = "Windows";
    private static final String OS_VERSION_WEB = "10";
    private static final String USER_AGENT_WEB = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0)" +
            " Gecko/20100101" +
            " Firefox/129.0";

    // ANDROID
    private static final String OS_NAME_ANDROID = "Android";

    // IOS
    /**
     * The hardcoded client version of the iOS app used for InnerTube requests with this client.
     *
     * <p>
     * It can be extracted by getting the latest release version of the app on
     * <a href="https://apps.apple.com/us/app/youtube-watch-listen-stream/id544007664/">the App
     * Store page of the YouTube app</a>, in the {@code What’s New} section.
     * </p>
     */
    private static final String CLIENT_VERSION_IOS = "19.49.5";
    private static final String DEVICE_MAKE_IOS = "Apple";
    /**
     * The device machine id for the iPhone 16 Pro Max (iPhone17,2), used to get HDR with AV1 hardware decoding.
     *
     * <p>
     * See <a href="https://gist.github.com/adamawolf/3048717">this GitHub Gist</a> for more
     * information.
     * </p>
     */
    private static final String DEVICE_MODEL_IOS = "iPhone17,2";
    private static final String OS_NAME_IOS = "iOS";
    /**
     * The minimum supported OS version for the iOS YouTube client is iOS 14.0.
     * Using an invalid OS version will use the AVC codec.
     */
    private static final String OS_VERSION_IOS = "18.1.1.22B91";
    private static final String USER_AGENT_VERSION_IOS = "18_1_1";
    private static final String USER_AGENT_IOS = "com.google.ios.youtube/" +
            CLIENT_VERSION_IOS +
            "(" +
            DEVICE_MODEL_IOS +
            "; U; CPU iOS " +
            USER_AGENT_VERSION_IOS +
            " like Mac OS X)";

    // IOS_MUSIC
    /**
     * The hardcoded client version of the iOS app used for InnerTube requests with this client.
     *
     * <p>
     * It can be extracted by getting the latest release version of the app on
     * <a href="https://apps.apple.com/us/app/youtube-music/id1017492454/">the App
     * Store page of the YouTube app</a>, in the {@code What’s New} section.
     * </p>
     */
    private static final String CLIENT_VERSION_IOS_MUSIC = "7.31.2";
    private static final String USER_AGENT_IOS_MUSIC = "com.google.ios.youtubemusic/" +
            CLIENT_VERSION_IOS +
            "(" +
            DEVICE_MODEL_IOS +
            "; U; CPU iOS " +
            USER_AGENT_VERSION_IOS +
            " like Mac OS X)";

    // ANDROID VR
    /**
     * The hardcoded client version of the Android VR app used for InnerTube requests with this client.
     *
     * <p>
     * It can be extracted by getting the latest release version of the app on
     * <a href="https://www.meta.com/en-us/experiences/2002317119880945/">the App
     * Store page of the YouTube app</a>, in the {@code Additional details} section.
     * </p>
     */
    private static final String CLIENT_VERSION_ANDROID_VR = "1.61.47";
    private static final String DEVICE_MAKE_ANDROID_VR = "Oculus";
    /**
     * The device machine id for the Meta Quest 3, used to get opus codec with the Android VR client.
     *
     * <p>
     * See <a href="https://dumps.tadiphone.dev/dumps/oculus/eureka">this GitLab</a> for more
     * information.
     * </p>
     */
    private static final String DEVICE_MODEL_ANDROID_VR = "Quest 3";
    private static final String OS_VERSION_ANDROID_VR = "12L";
    /**
     * The SDK version for Android 12 is 31,
     * but for some reason the build.props for the {@code Quest 3} state that the SDK version is 32.
     */
    private static final String ANDROID_SDK_VERSION_ANDROID_VR = "32";
    /**
     * Package name for YouTube VR (Google DayDream): com.google.android.apps.youtube.vr (Deprecated)
     * Package name for YouTube VR (Meta Quests): com.google.android.apps.youtube.vr.oculus
     * Package name for YouTube VR (ByteDance Pico 4): com.google.android.apps.youtube.vr.pico
     */
    private static final String USER_AGENT_ANDROID_VR = "com.google.android.apps.youtube.vr.oculus/" +
            CLIENT_VERSION_ANDROID_VR +
            " (Linux; U; Android " +
            OS_VERSION_ANDROID_VR +
            "; eureka-user Build/SQ3A.220605.009.A1) gzip";

    // ANDROID UNPLUGGED
    private static final String CLIENT_VERSION_ANDROID_UNPLUGGED = "8.49.0";
    /**
     * The device machine id for the Chromecast with Google TV 4K.
     *
     * <p>
     * See <a href="https://dumps.tadiphone.dev/dumps/google/kirkwood">this GitLab</a> for more
     * information.
     * </p>
     */
    private static final String DEVICE_MODEL_ANDROID_UNPLUGGED = "Google TV Streamer";
    private static final String OS_VERSION_ANDROID_UNPLUGGED = "14";
    private static final String ANDROID_SDK_VERSION_ANDROID_UNPLUGGED = "34";
    private static final String USER_AGENT_ANDROID_UNPLUGGED = "com.google.android.apps.youtube.unplugged/" +
            CLIENT_VERSION_ANDROID_UNPLUGGED +
            " (Linux; U; Android " +
            OS_VERSION_ANDROID_UNPLUGGED +
            "; GB) gzip";

    private AppClient() {
    }

    public enum ClientType {
        WEB(1,
                null,
                DEVICE_MODEL_WEB,
                OS_NAME_WEB,
                OS_VERSION_WEB,
                USER_AGENT_WEB,
                null,
                CLIENT_VERSION_WEB,
                true
        ),
        IOS(5,
                DEVICE_MAKE_IOS,
                DEVICE_MODEL_IOS,
                OS_NAME_IOS,
                OS_VERSION_IOS,
                USER_AGENT_IOS,
                null,
                CLIENT_VERSION_IOS,
                false
        ),
        ANDROID_VR(28,
                DEVICE_MAKE_ANDROID_VR,
                DEVICE_MODEL_ANDROID_VR,
                OS_NAME_ANDROID,
                OS_VERSION_ANDROID_VR,
                USER_AGENT_ANDROID_VR,
                ANDROID_SDK_VERSION_ANDROID_VR,
                CLIENT_VERSION_ANDROID_VR,
                true
        ),
        ANDROID_UNPLUGGED(29,
                null,
                DEVICE_MODEL_ANDROID_UNPLUGGED,
                OS_NAME_ANDROID,
                OS_VERSION_ANDROID_UNPLUGGED,
                USER_AGENT_ANDROID_UNPLUGGED,
                ANDROID_SDK_VERSION_ANDROID_UNPLUGGED,
                CLIENT_VERSION_ANDROID_UNPLUGGED,
                true
        ),
        IOS_MUSIC(26,
                DEVICE_MAKE_IOS,
                DEVICE_MODEL_IOS,
                OS_NAME_IOS,
                OS_VERSION_IOS,
                USER_AGENT_IOS_MUSIC,
                null,
                CLIENT_VERSION_IOS_MUSIC,
                true
        );

        /**
         * YouTube
         * <a href="https://github.com/zerodytrash/YouTube-Internal-Clients?tab=readme-ov-file#clients">client type</a>
         */
        public final int id;

        public final String clientName;

        /**
         * Device manufacturer.
         */
        @Nullable
        public final String deviceMake;

        /**
         * Device model, equivalent to {@link Build#MODEL} (System property: ro.product.model)
         */
        public final String deviceModel;

        /**
         * Device OS name.
         */
        public final String osName;

        /**
         * Device OS version.
         */
        public final String osVersion;

        /**
         * Player user-agent.
         */
        public final String userAgent;

        /**
         * Android SDK version, equivalent to {@link Build.VERSION#SDK} (System property: ro.build.version.sdk)
         * Field is null if not applicable.
         */
        @Nullable
        public final String androidSdkVersion;

        /**
         * App version.
         */
        public final String clientVersion;

        /**
         * If the client can access the API logged in.
         */
        public final boolean canLogin;

        ClientType(int id,
                   String deviceMake,
                   String deviceModel,
                   String osName,
                   String osVersion,
                   String userAgent,
                   @Nullable String androidSdkVersion,
                   String clientVersion,
                   boolean canLogin
        ) {
            this.id = id;
            this.clientName = name();
            this.deviceMake = deviceMake;
            this.deviceModel = deviceModel;
            this.clientVersion = clientVersion;
            this.osName = osName;
            this.osVersion = osVersion;
            this.androidSdkVersion = androidSdkVersion;
            this.userAgent = userAgent;
            this.canLogin = canLogin;
        }

        public final String getFriendlyName() {
            return getString("revanced_spoof_streaming_data_type_entry_" + name().toLowerCase());
        }
    }
}

File: extensions/shared/src/main/java/app/revanced/extension/shared/patches/client/AppClient.kt
package app.revanced.extension.shared.patches.client

import android.os.Build
import app.revanced.extension.shared.patches.PatchStatus
import app.revanced.extension.shared.settings.BaseSettings

object AppClient {
    // MWEB
    /**
     * Video not playable: None
     * Note: Audio track available
     */
    private const val CLIENT_VERSION_MWEB = "2.20241202.07.00";

    private const val DEVICE_MODEL_MWEB = "iPad";

    private const val OS_VERSION_MWEB = "16_7_10";

    /**
     * MWEB does not require PO Token with this User Agent
     */
    private const val USER_AGENT_MWEB =
            "Mozilla/5.0 (iPad; CPU OS 16_7_10 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1,gzip(gfe)";

    // IOS
    /**
     * Video not playable: Paid / Movie / Private / Age-restricted
     * Note: Audio track available
     */
    private const val PACKAGE_NAME_IOS = "com.google.ios.youtube"

    /**
     * The hardcoded client version of the iOS app used for InnerTube requests with this client.
     *
     * It can be extracted by getting the latest release version of the app on
     * [the App Store page of the YouTube app](https://apps.apple.com/us/app/youtube-watch-listen-stream/id544007664/),
     * in the `What’s New` section.
     */
    private val CLIENT_VERSION_IOS = if (forceAVC())
        "17.40.5"
    else
        "19.29.1"

    private const val DEVICE_MAKE_IOS = "Apple"

    /**
     * The device machine id for the iPhone 15 Pro Max (iPhone16,2),
     * used to get HDR with AV1 hardware decoding.
     * See [this GitHub Gist](https://gist.github.com/adamawolf/3048717) for more information.
     */
    private val DEVICE_MODEL_IOS = if (forceAVC())
        "iPhone12,5" // 11 Pro Max. (last device with iOS 13)
    else
        "iPhone16,2" // 15 Pro Max.
    private val OS_VERSION_IOS = if (forceAVC())
        "13.7.17H35" // Last release of iOS 13.
    else
        "17.7.2.21H221"
    private val USER_AGENT_VERSION_IOS = if (forceAVC())
        "13_7"
    else
        "17_7_2"
    private val USER_AGENT_IOS = iOSUserAgent(PACKAGE_NAME_IOS, CLIENT_VERSION_IOS)


    // IOS UNPLUGGED
    /**
     * Video not playable: Paid / Movie
     * Note: Audio track available
     */
    private const val PACKAGE_NAME_IOS_UNPLUGGED = "com.google.ios.youtubeunplugged"

    /**
     * The hardcoded client version of the iOS app used for InnerTube requests with this client.
     *
     * It can be extracted by getting the latest release version of the app on
     * [the App Store page of the YouTube TV app](https://apps.apple.com/us/app/youtube-tv/id1193350206/),
     * in the `What’s New` section.
     */
    private val CLIENT_VERSION_IOS_UNPLUGGED = if (forceAVC())
        "6.45"
    else
        "8.33"
    private val USER_AGENT_IOS_UNPLUGGED =
        iOSUserAgent(PACKAGE_NAME_IOS_UNPLUGGED, CLIENT_VERSION_IOS_UNPLUGGED)


    // ANDROID VR
    /**
     * Video not playable: Kids
     * Note: Audio track is not available
     *
     * Package name for YouTube VR (Google DayDream): com.google.android.apps.youtube.vr (Deprecated)
     * Package name for YouTube VR (Meta Quests): com.google.android.apps.youtube.vr.oculus
     * Package name for YouTube VR (ByteDance Pico 4): com.google.android.apps.youtube.vr.pico
     */
    private const val PACKAGE_NAME_ANDROID_VR = "com.google.android.apps.youtube.vr.oculus"

    /**
     * The hardcoded client version of the Android VR app used for InnerTube requests with this client.
     *
     * It can be extracted by getting the latest release version of the app on
     * [the App Store page of the YouTube app](https://www.meta.com/en-us/experiences/2002317119880945/),
     * in the `Additional details` section.
     */
    private const val CLIENT_VERSION_ANDROID_VR = "1.61.48"

    /**
     * The device machine id for the Meta Quest 3, used to get opus codec with the Android VR client.
     * See [this GitLab](https://dumps.tadiphone.dev/dumps/oculus/eureka) for more information.
     */
    private const val DEVICE_MAKE_ANDROID_VR = "Oculus"
    private const val DEVICE_MODEL_ANDROID_VR = "Quest 3"
    private const val OS_VERSION_ANDROID_VR = "12L"

    /**
     * The SDK version for Android 12 is 31,
     * but for some reason the build.props for the `Quest 3` state that the SDK version is 32.
     */
    private const val ANDROID_SDK_VERSION_ANDROID_VR = "32"
    private val USER_AGENT_ANDROID_VR =
        androidUserAgent(PACKAGE_NAME_ANDROID_VR, CLIENT_VERSION_ANDROID_VR, OS_VERSION_ANDROID_VR)


    // ANDROID UNPLUGGED
    /**
     * Video not playable: Playlists / Music
     * Note: Audio track is not available
     */
    private const val PACKAGE_NAME_ANDROID_UNPLUGGED = "com.google.android.apps.youtube.unplugged"
    private const val CLIENT_VERSION_ANDROID_UNPLUGGED = "8.16.0"

    /**
     * The device machine id for the Chromecast with Google TV 4K.
     * See [this GitLab](https://dumps.tadiphone.dev/dumps/google/kirkwood) for more information.
     */
    private const val DEVICE_MAKE_ANDROID_UNPLUGGED = "Google"
    private const val DEVICE_MODEL_ANDROID_UNPLUGGED = "Google TV Streamer"
    private const val OS_VERSION_ANDROID_UNPLUGGED = "14"
    private const val ANDROID_SDK_VERSION_ANDROID_UNPLUGGED = "34"
    private val USER_AGENT_ANDROID_UNPLUGGED = androidUserAgent(
        PACKAGE_NAME_ANDROID_UNPLUGGED,
        CLIENT_VERSION_ANDROID_UNPLUGGED,
        OS_VERSION_ANDROID_UNPLUGGED
    )


    // ANDROID MUSIC
    /**
     * Video not playable: All videos that can't be played on YouTube Music
     */
    private const val PACKAGE_NAME_ANDROID_MUSIC = "com.google.android.apps.youtube.music"

    /**
     * Older client versions don't seem to require poToken.
     * It is not the default client yet, as it requires sufficient testing.
     */
    private const val CLIENT_VERSION_ANDROID_MUSIC = "4.27.53"
    private val ANDROID_SDK_VERSION_ANDROID_MUSIC = Build.VERSION.SDK_INT.toString()
    private val USER_AGENT_ANDROID_MUSIC = androidUserAgent(
        PACKAGE_NAME_ANDROID_MUSIC,
        CLIENT_VERSION_ANDROID_MUSIC
    )


    private fun androidUserAgent(
        packageName: String,
        clientVersion: String,
        osVersion: String? = Build.VERSION.RELEASE
    ): String {
        return packageName +
                "/" +
                clientVersion +
                " (Linux; U; Android " +
                osVersion +
                "; GB) gzip"
    }

    private fun iOSUserAgent(packageName: String, clientVersion: String): String {
        return packageName +
                "/" +
                clientVersion +
                "(" +
                DEVICE_MODEL_IOS +
                "; U; CPU iOS " +
                USER_AGENT_VERSION_IOS +
                " like Mac OS X)"
    }

    private fun forceAVC(): Boolean {
        return BaseSettings.SPOOF_STREAMING_DATA_IOS_FORCE_AVC.get()
    }

    @JvmStatic
    val availableClientTypes: Array<ClientType>
        get() = if (PatchStatus.SpoofStreamingDataMusic())
            ClientType.CLIENT_ORDER_TO_USE_YOUTUBE_MUSIC
        else
            ClientType.CLIENT_ORDER_TO_USE_YOUTUBE

    enum class ClientType(
        /**
         * [YouTube client type](https://github.com/zerodytrash/YouTube-Internal-Clients?tab=readme-ov-file#clients)
         */
        val id: Int,
        /**
         * Device manufacturer, equivalent to [Build.MANUFACTURER] (System property: ro.product.manufacturer)
         */
        @JvmField
        val deviceMake: String? = Build.MANUFACTURER,

        /**
         * Device model, equivalent to [Build.MODEL] (System property: ro.product.model)
         */
        @JvmField
        val deviceModel: String? = Build.MODEL,
        /**
         * Device OS version, equivalent to [Build.VERSION.RELEASE] (System property: ro.system.build.version.release)
         */
        @JvmField
        val osVersion: String? = Build.VERSION.RELEASE,
        /**
         * Client user-agent.
         */
        @JvmField
        val userAgent: String,
        /**
         * Android SDK version, equivalent to [Build.VERSION.SDK] (System property: ro.build.version.sdk)
         * Field is null if not applicable.
         */
        @JvmField
        val androidSdkVersion: String? = null,
        /**
         * App version.
         */
        @JvmField
        val clientVersion: String,
        /**
         * If the client can access the API logged in.
         */
        @JvmField
        val canLogin: Boolean? = true,
        /**
         * If a poToken should be used.
         */
        @JvmField
        val usePoToken: Boolean? = false,
        /**
         * Friendly name displayed in stats for nerds.
         */
        @JvmField
        val friendlyName: String
    ) {
        MWEB(
            id = 2,
            deviceMake = DEVICE_MAKE_IOS,
            deviceModel = DEVICE_MODEL_MWEB,
            osVersion = OS_VERSION_MWEB,
            userAgent = USER_AGENT_MWEB,
            clientVersion = CLIENT_VERSION_MWEB,
            canLogin = false,
            friendlyName = "Mobile Web"
        ),
        ANDROID_VR(
            id = 28,
            deviceMake = DEVICE_MAKE_ANDROID_VR,
            deviceModel = DEVICE_MODEL_ANDROID_VR,
            osVersion = OS_VERSION_ANDROID_VR,
            userAgent = USER_AGENT_ANDROID_VR,
            androidSdkVersion = ANDROID_SDK_VERSION_ANDROID_VR,
            clientVersion = CLIENT_VERSION_ANDROID_VR,
            friendlyName = "Android VR"
        ),
        ANDROID_UNPLUGGED(
            id = 29,
            deviceMake = DEVICE_MAKE_ANDROID_UNPLUGGED,
            deviceModel = DEVICE_MODEL_ANDROID_UNPLUGGED,
            osVersion = OS_VERSION_ANDROID_UNPLUGGED,
            userAgent = USER_AGENT_ANDROID_UNPLUGGED,
            androidSdkVersion = ANDROID_SDK_VERSION_ANDROID_UNPLUGGED,
            clientVersion = CLIENT_VERSION_ANDROID_UNPLUGGED,
            friendlyName = "Android TV"
        ),
        IOS_UNPLUGGED(
            id = 33,
            deviceMake = DEVICE_MAKE_IOS,
            deviceModel = DEVICE_MODEL_IOS,
            osVersion = OS_VERSION_IOS,
            userAgent = USER_AGENT_IOS_UNPLUGGED,
            clientVersion = CLIENT_VERSION_IOS_UNPLUGGED,
            friendlyName = if (forceAVC())
                "iOS TV Force AVC"
            else
                "iOS TV"
        ),
        IOS(
            id = 5,
            deviceMake = DEVICE_MAKE_IOS,
            deviceModel = DEVICE_MODEL_IOS,
            osVersion = OS_VERSION_IOS,
            userAgent = USER_AGENT_IOS,
            clientVersion = CLIENT_VERSION_IOS,
            canLogin = false,
            usePoToken = true,
            friendlyName = if (forceAVC())
                "iOS Force AVC"
            else
                "iOS"
        ),
        ANDROID_MUSIC(
            id = 21,
            userAgent = USER_AGENT_ANDROID_MUSIC,
            androidSdkVersion = ANDROID_SDK_VERSION_ANDROID_MUSIC,
            clientVersion = CLIENT_VERSION_ANDROID_MUSIC,
            friendlyName = "Android Music"
        );

        @JvmField
        val clientName: String = name

        companion object {
            val CLIENT_ORDER_TO_USE_YOUTUBE: Array<ClientType> = arrayOf(
                ANDROID_VR,
                ANDROID_UNPLUGGED,
                ANDROID_MUSIC,
                IOS_UNPLUGGED,
                IOS,
            )

            internal val CLIENT_ORDER_TO_USE_YOUTUBE_MUSIC: Array<ClientType> = arrayOf(
                ANDROID_VR,
                ANDROID_MUSIC,
            )
        }
    }
}
